Dont increment generation if forward of backward exists, and warning on deallocation of live tensors

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97176
* #96775
* #97171
* __->__ #97168


Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

eellison committed:
Dont increment generation if forward of backward exists, and warning on deallocation of live tensors

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont assume increment generation. 

Also adds warnings. 


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44228369).

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -1083,7 +1111,18 @@ def __init__(self, device_index: int):
+        self.id_to_mode: Dict[int, CompilationMode] = {}
+
+        # forwards that have been invoked without invocation of their corresponding backwards
+        self.called_forwards_on_stack: int = 0
```
can we call this something more meaningful, like `forwards_with_pending_backwards`

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -162,6 +162,11 @@ def compile_fx_inner(
+
+    graph_name = torch._functorch.aot_autograd.get_graph_being_compiled()
+    graph_no_id_suffix = graph_name[0 : -(len(str(graph_id)) + 1)]
+    is_inference = graph_no_id_suffix[-(len("inference")) :] == "inference"
```
No. Figure out a better way to do this

eellison replied:
It's a breaking change for all of our backends 😩 but sure

eellison replied:
I think I'll have aot autograd take in an inference backend and default it to fwd_compiler, if that makes sense to you ? cc @Chillee

ezyang replied:
I think you should just stuff it in TracingContext in the (somewhat misnamed) torch._guards

ezyang replied:
cc @voznesenskym

eellison replied:
Adapted an inference_compiler option.

ezyang reviewed (dismissed):
The substantive logic is fine, but you should figure out another way to determine if something is inference or not.

The easiest is probably to have AOTAutograd transmit this information to the backend.

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

eellison committed:
Update on "Dont increment generation if forward of backward exists, and warning on deallocation of live tensors"



Refining the logic for when it is okay to ignore previously live outputs from cudagraphs. If there is a forward that has been invoked without invocation of the corresponding backwards, dont allow overwriting outputs.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228369](https://our.internmc.facebook.com/intern/diff/D44228369)

[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2765,7 +2769,10 @@ def aot_function(
             larger Aten ops into simpler or core Aten ops.
-
+        inference_compiler (Optional[Callable]): A Python function that accepts an
+            Fx graph with Aten ops and input args, and returns a Callable that
```
A better explanation of the difference between forward and inference would help

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -183,6 +185,8 @@ def compile_fx_inner(
                 return compiled_fn
 
+    output = list(gm.graph.nodes)[-1]
+    assert len(output.args) == 1
```
?

eellison replied:
I was a bit surprised that the "output" node has a single arg with tuple inputs, instead of multiple args. I added an assert to verify that. will add comment

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -1196,11 +1235,27 @@ def new_graph_id(self) -> GraphID:
+        inputs,
+        static_input_idxs,
+        is_backward,
+        is_inference,
```
Why doesn't this use the enum? As is you have the impossible state is backward AND is inference

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -1083,7 +1111,18 @@ def __init__(self, device_index: int):
+
     def run(self, new_inputs: List[Tensor], function_id: FunctionID):
+        mode = self.id_to_mode[function_id]
+        if mode == CompilationMode.FORWARD:
```
Nit: use is instead of equality, it is better for mypy

eellison commented:
@pytorchbot merge
